### PR TITLE
Strategize slicing based on free [V]RAM

### PR DESCRIPTION
**Pull request by JPPhoto**
I had some spare time so I strategized slicing - none or `max` - at runtime based on the size of the generation requested and free [V]RAM at that time. This needs testers on multiple platforms.

### CODE REVIEW

**Comment by Kyle0654** [comment]
I know we're nowhere near there, but do you have any ideas about how we'd make this work in a batched/parallel-generation environment?

**Comment by JPPhoto** [comment]
> I know we're nowhere near there, but do you have any ideas about how we'd make this work in a batched/parallel-generation environment?

We're nowhere near there. :)

For a batched system, each batch gets to make the call as to whether there's available [V]RAM to run fast or `max` sliced. I don't think there are any code changes required for that case.

If we're on a parallel-generation system we'd have to have some awareness of the other concurrent generations on the system and add all of their memory requirements first before deciding how to run ours, or (the simple solution) every job on that type of system has to run with `max` to avoid OOM errors. I also imagine that sub-quadratic slicing will take care of a lot of these issues when it's implemented in Invoke or (better) diffusers.

**Comment by JPPhoto** [comment]
I also want to get some Windows + CUDA testers to run in a debugger and see if you're getting an accurate amount of free VRAM back from torch. If not, everything will still run - just in slices in all cases.

**Comment by lstein** [comment]
Thanks for doing this. I am looking forward to testing this feature after the 2.3.0 dust settles

**Comment by psychedelicious** [comment]
@JPPhoto I'm happy to test, but I'm not sure what I'm looking for... should I just see if the magic smoke escapes?

**Comment by JPPhoto** [comment]
> @JPPhoto I'm happy to test, but I'm not sure what I'm looking for... should I just see if the magic smoke escapes?

That is definitely part of it. But I want to make sure you can render at sizes that are appropriate for your free [V]RAM without crashing with as much speed as possible. What's your setup?

On my 12GB NVIDIA card, tensors for a 512x512 image fit entirely in VRAM alongside the model. When I scale up, memory requirements grow - the rough formula is that you need 16 * ((x * y / 64) ^ 2) * 6 (or 8 if using fp32) bytes of [V]RAM. So a 1280x1280 image is possible on my card via slicing implemented by diffusers and shouldn't cause an OOM.

**Comment by psychedelicious** [comment]
I'm on an M1 MacBook with 32GB memory (shared RAM/VRAM). I don't think invoke can access my VRAM usage, so anything that makes decisions based on that probably will not work.

**Review by lstein — APPROVED** [review]
Confirmed working on Ubuntu.

File: ldm/invoke/generator/diffusers_pipeline.py [hunk 1]
```diff
@@ -377,6 +392,7 @@ def generate_latents_from_embeddings(self, latents: torch.Tensor, timesteps,
                                          noise: torch.Tensor,
                                          run_id: str = None,
                                          additional_guidance: List[Callable] = None):
+        self._enable_memory_efficient_attention(latents)
```
Comment (damian0815) [inline ldm/invoke/generator/diffusers_pipeline.py:395]: if we're calling this every generate, it should be called not `_enable` but idk `_adjust_memory_efficient_attention` or something like that.

**Review by damian0815 — APPROVED** [review]
looks good other than the naming the `_enable_memory_efficient_attention` function

File: ldm/invoke/generator/diffusers_pipeline.py (hunk 1 above)
Comment (JPPhoto) [reply ldm/invoke/generator/diffusers_pipeline.py]: Fixed.

**Pull request merged**